File: library/src/main/java/com/yy/codex/uikit/UIControl.java
package com.yy.codex.uikit;

import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.view.View;

import com.yy.codex.foundation.NSInvocation;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by cuiminghui on 2017/1/13.
 */

public class UIControl extends UIView {

    public enum Event {
        TouchDown,
        TouchDownRepeat,
        TouchDragInside,
        TouchDragOutside,
        TouchDragEnter,
        TouchDragExit,
        TouchUpInside,
        TouchUpOutside,
        TouchCancel,
        ValueChanged
    }

    public enum State {
        Normal,
        Highlighted,
        Disabled,
        Selected,
    }

    public enum ContentVerticalAlignment {
        Center,
        Top,
        Bottom,
        Fill,
    }

    public enum ContentHorizontalAlignment {
        Center,
        Left,
        Right,
        Fill,
    }

    public UIControl(@NonNull Context context, @NonNull View view) {
        super(context, view);
    }

    public UIControl(@NonNull Context context) {
        super(context);
    }

    public UIControl(@NonNull Context context, @NonNull AttributeSet attrs) {
        super(context, attrs);
    }

    public UIControl(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public UIControl(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    private HashMap<Event, NSInvocation[]> mInvocations = new HashMap<>();
    private HashMap<Event, Runnable[]> mRunnable = new HashMap<>();

    public void addTarget(@NonNull Object target, @NonNull String selector, @NonNull Event event) {
        if (!mInvocations.containsKey(event)) {
            mInvocations.put(event, new NSInvocation[0]);
        }
        NSInvocation[] invocations = new NSInvocation[mInvocations.get(event).length + 1];
        for (int i = 0; i < mInvocations.get(event).length; i++) {
            invocations[i] = mInvocations.get(event)[i];
        }
        invocations[invocations.length - 1] = new NSInvocation(target, selector);
        mInvocations.put(event, invocations);
    }

    public void removeTarget(@Nullable Object target, @Nullable String selector, @Nullable Event event) {
        if (target == null && selector == null && event == null) {
            mInvocations.clear();
        }
        else if (target == null && selector == null && event != null) {
            mInvocations.remove(event);
        }
        else if (target != null && selector == null && event == null) {
            for (Map.Entry<Event, NSInvocation[]> entry: mInvocations.entrySet()) {
                NSInvocation[] invocations = entry.getValue();
                ArrayList<NSInvocation> arrayList = new ArrayList<>();
                for (int i = 0; i < invocations.length; i++) {
                    if (invocations[i].getTarget() != target) {
                        arrayList.add(invocations[i]);
                    }
                }
                NSInvocation[] newInvocations = new NSInvocation[arrayList.size()];
                arrayList.toArray(newInvocations);
                mInvocations.put(entry.getKey(), newInvocations);
            }
        }
        else if (target != null && selector != null && event == null) {
            for (Map.Entry<Event, NSInvocation[]> entry: mInvocations.entrySet()) {
                NSInvocation[] invocations = entry.getValue();
                ArrayList<NSInvocation> arrayList = new ArrayList<>();
                for (int i = 0; i < invocations.length; i++) {
                    if (invocations[i].getTarget() != target || invocations[i].getSelector() != selector) {
                        arrayList.add(invocations[i]);
                    }
                }
                NSInvocation[] newInvocations = new NSInvocation[arrayList.size()];
                arrayList.toArray(newInvocations);
                mInvocations.put(entry.getKey(), newInvocations);
            }
        }
        else if (target == null && selector != null && event == null) {
            for (Map.Entry<Event, NSInvocation[]> entry: mInvocations.entrySet()) {
                NSInvocation[] invocations = entry.getValue();
                ArrayList<NSInvocation> arrayList = new ArrayList<>();
                for (int i = 0; i < invocations.length; i++) {
                    if (invocations[i].getSelector() != selector) {
                        arrayList.add(invocations[i]);
                    }
                }
                NSInvocation[] newInvocations = new NSInvocation[arrayList.size()];
                arrayList.toArray(newInvocations);
                mInvocations.put(entry.getKey(), newInvocations);
            }
        }
        else if (target == null && selector != null && event != null) {
            if (!mInvocations.containsKey(event)) {
                return;
            }
            NSInvocation[] invocations = mInvocations.get(event);
            ArrayList<NSInvocation> arrayList = new ArrayList<>();
            for (int i = 0; i < invocations.length; i++) {
                if (invocations[i].getTarget() != target) {
                    arrayList.add(invocations[i]);
                }
            }
            NSInvocation[] newInvocations = new NSInvocation[arrayList.size()];
            arrayList.toArray(newInvocations);
            mInvocations.put(event, newInvocations);
        }
        else if (target != null && selector != null && event != null) {
            if (!mInvocations.containsKey(event)) {
                return;
            }
            NSInvocation[] invocations = mInvocations.get(event);
            ArrayList<NSInvocation> arrayList = new ArrayList<>();
            for (int i = 0; i < invocations.length; i++) {
                if (invocations[i].getTarget() != target || invocations[i].getSelector() != selector) {
                    arrayList.add(invocations[i]);
                }
            }
            NSInvocation[] newInvocations = new NSInvocation[arrayList.size()];
            arrayList.toArray(newInvocations);
            mInvocations.put(event, newInvocations);
        }
    }

    public void addBlock(@NonNull Runnable runnable,@NonNull Event event) {
        if (!mRunnable.containsKey(event)) {
            mRunnable.put(event, new Runnable[0]);
        }
        Runnable[] runnables = new Runnable[mRunnable.get(event).length + 1];
        for (int i = 0; i < mRunnable.get(event).length; i++) {
            runnables[i] = mRunnable.get(event)[i];
        }
        runnables[runnables.length - 1] = runnable;
        mRunnable.put(event, runnables);
    }

    @Override
    protected void init() {
        super.init();
        setUserInteractionEnabled(true);
        UILongPressGestureRecognizer longPressGestureRecognizer = new UILongPressGestureRecognizer(this, "onLongPressed:");
        longPressGestureRecognizer.mMinimumPressDuration = 0.05;
        addGestureRecognizer(longPressGestureRecognizer);
        UITapGestureRecognizer tapGestureRecognizer = new UITapGestureRecognizer(this, "onTapped:");
        addGestureRecognizer(tapGestureRecognizer);
    }

    private boolean mInside = true;

    protected void onLongPressed(UILongPressGestureRecognizer sender) {
        if (sender.mState == UIGestureRecognizerState.Began) {
            mInside = true;
            onEvent(Event.TouchDown);
            mTracking = true;
            mHighlighted = true;
            resetState();
        }
        else if (sender.mState == UIGestureRecognizerState.Changed) {
            if (isPointInside(sender.location(this))) {
                onEvent(Event.TouchDragInside);
                if (!mInside) {
                    onEvent(Event.TouchDragEnter);
                    mInside = true;
                    mHighlighted = true;
                    resetState();
                }
            }
            else {
                onEvent(Event.TouchDragOutside);
                if (mInside) {
                    onEvent(Event.TouchDragExit);
                    mInside = false;
                    mHighlighted = false;
                    resetState();
                }
            }
        }
        else if (sender.mState == UIGestureRecognizerState.Ended) {
            if (isPointInside(sender.location(this))) {
                onEvent(Event.TouchUpInside);
            }
            else {
                onEvent(Event.TouchUpOutside);
            }
            mTracking = false;
            mHighlighted = false;
            resetState();
        }
    }

    protected void onTapped(UITapGestureRecognizer sender) {
        onEvent(Event.TouchUpInside);
    }

    protected void onEvent(Event event) {
        NSInvocation[] invocations = mInvocations.get(event);
        if (invocations != null) {
            for (int i = 0; i < invocations.length; i++) {
                invocations[i].invoke(new Object[]{this});
            }
        }
        Runnable[] runnables = mRunnable.get(event);
        if (runnables != null) {
            for (int i = 0; i < runnables.length; i++) {
                runnables[i].run();
            }
        }
    }

    protected boolean isPointInside(CGPoint point) {
        double xRange = this.getFrame().getSize().getWidth() / 2.0;
        double yRange = this.getFrame().getSize().getHeight() / 2.0;
        return point.inRange(xRange + 22, yRange + 22, new CGPoint(xRange, yRange));
    }

    /* Enabled */

    private boolean mEnabled = true;

    @Override
    public void setEnabled(boolean enabled) {
        this.mEnabled = enabled;
        setUserInteractionEnabled(enabled);
        resetState();
    }

    @Override
    public boolean isEnabled() {
        return mEnabled;
    }

    /* Selected */

    private boolean mSelected = false;

    @Override
    public void setSelected(boolean selected) {
        this.mSelected = selected;
        resetState();
    }

    @Override
    public boolean isSelected() {
        return mSelected;
    }

    /* Highlighted */

    private boolean mHighlighted = false;

    public boolean isHighlighted() {
        return mHighlighted;
    }

    /* State */

    private EnumSet<State> mState = EnumSet.of(State.Normal);

    protected void resetState() {
        EnumSet<State> state = EnumSet.of(State.Normal);
        if (isTracking() && isHighlighted()) {
            state.add(State.Highlighted);
        }
        if (isSelected()) {
            state.remove(State.Normal);
            state.add(State.Selected);
        }
        if (!isEnabled()) {
            state.add(State.Disabled);
        }
        mState = state;
    }

    public EnumSet<State> getState() {
        return mState;
    }

    /* Tracking */

    private boolean mTracking = false;

    public boolean isTracking() {
        return mTracking;
    }

    public boolean isTouchInside() {
        return mInside;
    }

    /* ContentAlignment */

    private ContentVerticalAlignment mContentVerticalAlignment = ContentVerticalAlignment.Center;

    public void setContentVerticalAlignment(ContentVerticalAlignment contentVerticalAlignment) {
        this.mContentVerticalAlignment = contentVerticalAlignment;
    }

    public ContentVerticalAlignment getContentVerticalAlignment() {
        return mContentVerticalAlignment;
    }

    private ContentHorizontalAlignment mContentHorizontalAlignment = ContentHorizontalAlignment.Center;

    public void setContentHorizontalAlignment(ContentHorizontalAlignment contentHorizontalAlignment) {
        this.mContentHorizontalAlignment = contentHorizontalAlignment;
    }

    public ContentHorizontalAlignment getContentHorizontalAlignment() {
        return mContentHorizontalAlignment;
    }

}

File: library/src/main/java/com/yy/codex/uikit/UIControl.kt
package com.yy.codex.uikit

import android.content.Context
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.view.View

import com.yy.codex.foundation.NSInvocation

import java.util.ArrayList
import java.util.EnumSet
import java.util.HashMap

/**
 * Created by cuiminghui on 2017/1/13.
 */

open class UIControl : UIView {

    enum class Event {
        TouchDown,
        TouchDownRepeat,
        TouchDragInside,
        TouchDragOutside,
        TouchDragEnter,
        TouchDragExit,
        TouchUpInside,
        TouchUpOutside,
        TouchCancel,
        ValueChanged
    }

    enum class State {
        Normal,
        Highlighted,
        Disabled,
        Selected
    }

    enum class ContentVerticalAlignment {
        Center,
        Top,
        Bottom,
        Fill
    }

    enum class ContentHorizontalAlignment {
        Center,
        Left,
        Right,
        Fill
    }

    constructor(context: Context, view: View) : super(context, view) {}

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {}

    private val invocations = HashMap<Event, List<NSInvocation>>()
    private val runnable = HashMap<Event, List<Runnable>>()

    override fun prepareProps(attrs: AttributeSet) {
        initializeAttributes = context.theme.obtainStyledAttributes(attrs, R.styleable.UIControl, 0, 0)
    }

    override fun resetProps() {
        super.resetProps()
        initializeAttributes?.let {
            enable = it.getBoolean(R.styleable.UIControl_enabled, true)
            select = it.getBoolean(R.styleable.UIControl_selected, false)
            when (it.getInt(R.styleable.UIControl_contentVerticalAlignment, 0)) {
                0 -> contentVerticalAlignment = ContentVerticalAlignment.Center
                1 -> contentVerticalAlignment = ContentVerticalAlignment.Top
                2 -> contentVerticalAlignment = ContentVerticalAlignment.Bottom
                3 -> contentVerticalAlignment = ContentVerticalAlignment.Fill
            }
            when (it.getInt(R.styleable.UIControl_contentHorizontalAlignment, 0)) {
                0 -> contentHorizontalAlignment = ContentHorizontalAlignment.Center
                1 -> contentHorizontalAlignment = ContentHorizontalAlignment.Left
                2 -> contentHorizontalAlignment = ContentHorizontalAlignment.Right
                3 -> contentHorizontalAlignment = ContentHorizontalAlignment.Fill
            }
        }
    }

    fun addTarget(target: Any, selector: String, event: Event) {
        if (!invocations.containsKey(event)) {
            invocations.put(event, listOf())
        }
        invocations[event]?.let {
            val mutableList = it.toMutableList()
            mutableList.add(NSInvocation(target, selector))
            invocations.put(event, mutableList.toList())
        }
    }

    fun removeTarget(target: Any?, selector: String?, event: Event?) {
        if (target == null && selector == null && event == null) {
            invocations.clear()
        } else if (target == null && selector == null && event != null) {
            invocations.remove(event)
        } else if (target != null && selector == null && event == null) {
            for ((key, invocations) in invocations) {
                this.invocations.put(key, invocations.filter { it.target !== target })
            }
        } else if (target != null && selector != null && event == null) {
            for ((key, invocations) in invocations) {
                this.invocations.put(key, invocations.filter { it.target !== target || it.selector !== selector })
            }
        } else if (target == null && selector != null && event == null) {
            for ((key, invocations) in invocations) {
                this.invocations.put(key, invocations.filter { it.selector !== selector })
            }
        } else if (target == null && selector != null && event != null) {
            if (!invocations.containsKey(event)) {
                return
            }
            val invocations = invocations[event]
            invocations?.let {
                this.invocations.put(event, invocations.filter { it.target !== target })
            }
        } else if (target != null && selector != null && event != null) {
            if (!invocations.containsKey(event)) {
                return
            }
            val invocations = invocations[event]
            invocations?.let {
                this.invocations.put(event, invocations.filter { it.target !== target || it.selector !== selector })
            }
        }
    }

    fun addBlock(runnable: Runnable, event: Event) {
        val allRunnable: MutableList<Runnable> = this.runnable[event]?.toMutableList() ?: mutableListOf()
        allRunnable.add(runnable)
        this.runnable.put(event, allRunnable.toList())
    }

    override fun init() {
        super.init()
        userInteractionEnabled = true
        val longPressGestureRecognizer = UILongPressGestureRecognizer(this, "onLongPressed:")
        longPressGestureRecognizer.minimumPressDuration = 0.05
        addGestureRecognizer(longPressGestureRecognizer)
        val tapGestureRecognizer = UITapGestureRecognizer(this, "onTapped:")
        addGestureRecognizer(tapGestureRecognizer)
    }

    var touchingInside = true
        private set

    protected open fun onLongPressed(sender: UILongPressGestureRecognizer) {
        if (sender.state == UIGestureRecognizerState.Began) {
            touchingInside = true
            onEvent(Event.TouchDown)
            tracking = true
            highlighted = true
            resetState()
        } else if (sender.state == UIGestureRecognizerState.Changed) {
            if (isPointInside(sender.location(this))) {
                onEvent(Event.TouchDragInside)
                if (!touchingInside) {
                    onEvent(Event.TouchDragEnter)
                    touchingInside = true
                    highlighted = true
                    resetState()
                }
            } else {
                onEvent(Event.TouchDragOutside)
                if (touchingInside) {
                    onEvent(Event.TouchDragExit)
                    touchingInside = false
                    highlighted = false
                    resetState()
                }
            }
        } else if (sender.state == UIGestureRecognizerState.Ended) {
            if (isPointInside(sender.location(this))) {
                onEvent(Event.TouchUpInside)
            } else {
                onEvent(Event.TouchUpOutside)
            }
            tracking = false
            highlighted = false
            resetState()
        }
    }

    protected fun onTapped(sender: UITapGestureRecognizer) {
        onEvent(Event.TouchUpInside)
    }

    protected open fun onEvent(event: Event) {
        val invocations = invocations[event]
        if (invocations != null) {
            for (i in invocations.indices) {
                invocations[i].invoke(arrayOf<Any>(this))
            }
        }
        val runnables = runnable[event]
        if (runnables != null) {
            for (i in runnables.indices) {
                runnables[i].run()
            }
        }
    }

    protected fun isPointInside(point: CGPoint): Boolean {
        val xRange = this.frame.size.width / 2.0
        val yRange = this.frame.size.height / 2.0
        return point.inRange(xRange + 22, yRange + 22, CGPoint(xRange, yRange))
    }

    var enable = true
        set(value) {
            field = value
            userInteractionEnabled = enable
            resetState()
        }

    var select = false
        set(value) {
            field = value
            resetState()
        }

    var highlighted = false
        private set

    var tracking = false
        private set

    var state: EnumSet<State> = EnumSet.of(State.Normal)
        private set

    protected open fun resetState() {
        val state = EnumSet.of(State.Normal)
        if (tracking && highlighted) {
            state.add(State.Highlighted)
        }
        if (select) {
            state.remove(State.Normal)
            state.add(State.Selected)
        }
        if (!enable) {
            state.add(State.Disabled)
        }
        this.state = state
    }

    var contentVerticalAlignment = ContentVerticalAlignment.Center

    var contentHorizontalAlignment = ContentHorizontalAlignment.Center

}
